### DISCUSSION TITLE
Change datasource for a single package

### DISCUSSION THREAD

**Discussion by Xerkus** [renovatebot/renovate#25018] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I am looking for a way to change datasource for a single package.

Specifically I want to make `php` with composer manager to look up tags on php/php-src again to pull release candidate versions. It is a special case with composer manager setting datasource to github tags but using containerbase/php-prebuild as package name which does not have unstable tags.

As a framework we want to introduce support for upcoming php release before stable is GA.
I was planning to change extract version regex to strip RC suffix since php-src tags do not use format for RC that renovate understands.
However unless I can change package name used for lookup or manually specify datasource and its config for this dependency I have no such option. I tried "packageName" but config validation says migration is required and changes it to "matchPackageName" which suggests it is not what I want.

Custom datasource with custom regex manager is not a suitable alternative here.

I do not know inner workings of renovate to look up how to do it in the source if it is at all possible.

### What you tried so far.

```json
    "packageRules": [
        {
            "matchPackageNames": ["php"],
            "extractVersion": "^php-(?<version>.*?)(RC[0-9])?$",
            "packageName": "php/php-src"
        }
    ],
```

**Comment by rarkins** [comment]
It's not currently possible, but in theory could be added. We may want to allow the user to rewrite not only datasource but also depName and packageName. Two challenges:
- With this power, clearly you could easily shoot yourself in the foot. We need to warn users
- Do we call these fields `datasource`, `depName`, and `packageName` or do we instead make it clearer using `setDatasource`, `setDepName`, and `setPackageName`

@viceice @secustor @HonkingGoose WDYT?

**Reply by HonkingGoose** [reply]
We should definitely warn users to be careful!

I like `setDatasource`, `setDepName` and `setPackageName`. Whatever names you pick, it should be clear what's "default Renovate internal `datasource`, `depName` and `packageName` stuff" and what's user created.

**Reply by Xerkus** [reply]
Changing depName is more or less equivalent to using replacement is it not?

If names with a verb to be used I would go for the likes of `overrideDatasource` or `forceDatasource` and `overridePackageName`/`overridePackageNameTemplate` to clearly indicate they are not needed and force different selection.

Such forced datasources are of very limited utility when it comes to composer. Composer needs repository entries in composer.json to override the package source and composer manager already does that.
I can think of just one dependency where it could be somewhat useful - `ext-swoole` or `ext-openswoole`. But then again it is not a true dependency that composer manages but rather a platform constraint.

There are some cases I can think of for other managers where it might be desirable to override datasource. Like docker when pull through caching registry is used. Eg image pointing at private AWS ECR pull-through registry but using original quay repository as datasource. Azure container registry might land such feature sometime soon-ish, I've seen it on their roadmap.

**Reply by rarkins** [reply]
I had in mind depName because it's used as part of the PR title and the branchName, but it could indeed cause chaos if we're not paying attention when we update. We _don't_ want to use this as another way to do replaces.

And good point re: force or override - it indicates the intention more

**Reply by Xerkus** [reply]
Might be better with `overrideDatasourcePackageName` to indicate what it is used for but it gets a bit too handful heh.

**Reply by secustor** [reply]
> And good point re: force or override - it indicates the intention more

I prefer not to add them as it diviates from the notion we mostly follow currently.
Which is that everything besides `match` and `exclude` prefixed values are properties.

We use `force` in the bot config too, which can lead to confusion while searching the docs. 

If we can make the generic overwrite work, it would IMO bring a lot of additional flexibility to packageRules

**Reply by Xerkus** [reply]
@secustor  I think using just `packageName` will conflict with previous configuration versions since config validator suggests migration for it to be `"matchPackageNames": [...],`

**Reply by HonkingGoose** [reply]
> With this power, clearly you could easily shoot yourself in the foot. We need to warn users

Even better, we should _automate_ the warning. 😉 You could probably copy/paste and adjust parts of:

https://github.com/renovatebot/renovate/blob/6aaf8df9f2aab12ad1124ca5dd52f3dac1a65f07/tools/docs/config.ts#L157-L180

And then add a new field to the config option object, similar to `experimental: true,` and `experimentalDescription`:

https://github.com/renovatebot/renovate/blob/6aaf8df9f2aab12ad1124ca5dd52f3dac1a65f07/lib/config/options/index.ts#L135-L145

---

**Edit:** turns out we _already_ automated this warning. 🙃 We have the `advancedUse` flag, which the docs code uses here:

https://github.com/renovatebot/renovate/blob/3c0f71d1e31e7ed16153ac2e338b46986e2642ec/tools/docs/config.ts#L237-L254

**Reply by viceice** [reply]
I like the general idea

**Reply by HonkingGoose** [reply]
@viceice we already have the `advancedUse` flag and docs generating code, I edited my comment above to show the code... 🙃

**Comment by Xerkus** [comment]
What I want to achieve here might be solved by a new generic internal feature - composite datasource: allow managers to configure several datasources with results getting combined. Eg lookup releases in public channel and unstable previews on another.

In case of php look up releases via whatever source is more sensible and unstable via php-src tags.

Limited insight I gained into renovate inner workings today indicates it would be quite extensive change for a questionable returns. May be it is something that would be useful with other managers so I mention it anyway.

**Reply by rarkins** [reply]
That doesn't sound like multiple datasources - in Renovate terminology it's multiple registryUrls

**Reply by Xerkus** [reply]
I've seen merge strategy that uses registry urls but it uses exactly the same config and datasource module with the only difference as registry url. So same package name, same extract version regex and so on.
Eg for github-tags it would choosing between github.com and GHE domains.

I meant possibility to combine distinct datasource modules or datasource configs. Different package name, different extract version regex, different type of source etc. 
I thought about it some and I do not like this suggestion at all. No sensible dependency tool would need that and those that might will just implement it as custom datasource.

**Comment by rarkins** [comment, marked as answer]
This topic has diverged from what the OP apparently meant, but I like the idea anyway so have created a Feature request issue for future comments to be moved to: #25053
